### PR TITLE
APEXMALHAR-2019 S3-Input Implemented S3 Input Module

### PR DESCRIPTION
**Pull request by chaithu14**

### CODE REVIEW

File: library/src/main/java/com/datatorrent/lib/io/fs/S3BlockReader.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    super.setup(context);
+    s3bucketUri = fs.getScheme() + "://" + extractBucket(uri);
+  }
+
+  @VisibleForTesting
+  protected String extractBucket(String s3uri)
```
Comment (yogidevendra) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3BlockReader.java]: javadoc.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3BlockReader.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  /**
+   * BlockReadeContext for reading S3 Blocks.<br/>
+   * This should use read API without offset.
+   */
+  private static class S3BlockReaderContext extends ReaderContext.FixedBytesReaderContext<FSDataInputStream>
+  {
+    @Override
+    protected Entity readEntity() throws IOException
```
Comment (yogidevendra) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3BlockReader.java:91]: mention the difference from super in javadocs.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3FileSplitter.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * S3FileSplitter extends from FSFileSplitter and dissables the parallel read functionality
+ * explicitly in case of s3 or s3n schemes.
+ * Parallel read will work only if the scheme is "s3a" and the Hadoop version is 2.7+.
+ * Parallel read doesn't work in the case of the scheme is "s3n/s3". In this case, this operator explicitly
+ * disables the parallel read functionality.
+ */
+public class S3FileSplitter extends FSFileSplitter
+{
+  public S3FileSplitter()
+  {
```
Comment (yogidevendra) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3FileSplitter.java]: Not required.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3FileSplitter.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+
+import java.io.IOException;
+import org.apache.hadoop.fs.FileSystem;
+import com.datatorrent.api.Context;
+
+/**
+ * S3FileSplitter extends from FSFileSplitter and dissables the parallel read functionality
+ * explicitly in case of s3 or s3n schemes.
+ * Parallel read will work only if the scheme is "s3a" and the Hadoop version is 2.7+.
```
Comment (yogidevendra) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3FileSplitter.java]: add link to amazon s3 wiki page to explain details about hadoop version and scheme.

**Comment by chaithu14** [comment]
Addressed all the review comments.

**Comment by gauravgopi123** [comment]
I see that this is using BlockReader, which is being deleted in PR https://github.com/apache/incubator-apex-malhar/pull/267. Please make sure that this is in line with other PR. So please hold off this till that time

**Comment by chaithu14** [comment]
Re-based with the master and updated the branch.

**Comment by chaithu14** [comment]
@ashwinchandrap Please review and merge.

**Comment by gauravgopi123** [comment]
Can we add test cases for this?

**Comment by chaithu14** [comment]
@gauravgopi123 Thanks for reviewing. I added unit test case for S3 Input Module. But, I marked the test case as excluded. Because to run this test case, the prerequisites are amazon credentials and hadoop.

File: library/src/main/java/com/datatorrent/lib/io/fs/S3BlockReader.java (hunk 2 above)
Comment (yogidevendra) [inline library/src/main/java/com/datatorrent/lib/io/fs/S3BlockReader.java:91]: @chaithu14 Could you please add javadoc to readEntity() mentioning the difference between the super implementation?

**Comment by yogidevendra** [comment]
@chaithu14 Please make changes to the readEntity() javadoc and please rebase this branch. 
If there are no comments from other community members then I will go ahead and merge this.

**Comment by yogidevendra** [comment]
@chaithu14 Could you please mark the new classes with @ evolving?

**Comment by chaithu14** [comment]
@yogidevendra : Incorporated review comments and marked the new classes with @ evolving.

**Comment by yogidevendra** [comment]
Looks fine to me. I will merge this tomorrow.

**Pull request merged**